### ISSUE TITLE
vimtex fails to load

### ISSUE THREAD

**Issue by brianbuccola** [lervag/vimtex#586]
(This is probably the same issue as [here](https://github.com/lervag/vimtex/issues/411), but I didn't want to necrobump a closed issue.)

### vimtex fails to load

**Problem.** vimtex "loads" in the sense that it's added to `'runtimepath'`; however, none of the commands, mappings, etc. are available. Only the documentation (e.g. `:h vimtex`) is available. Note that all other plugins (including LaTeX-Box, before I switched) work just fine, but even when I remove all plugins, vimtex still doesn't work.

**Environment.** I am using vim8's native package management system: all plugins that should be autoloaded go in `~/.vim/pack/my-plugins/start/`, and all plugins that are optionally loaded go in `~/.vim/pack/my-plugins/opt/` and are manually added with `:packadd`. The issue persists regardless of whether vimtex is autoloaded in `start/` or manually loaded from `opt/` with `:packadd`. For the illustration below, let's assume vimtex is the only plugin in `start/`.

**Steps to reproduce.**

1. Open minimal working example using minimal vimrc.
2. Expected behavior: vimtex commands, mappings, etc. should all be available.
3. Observed behaviour: they're not available; only the documentation is (i.e. `:h vimtex` works).

Note that `set filetype?` correctly shows `filetype=tex`, and `set 'runtimepath'?` shows that vimtex is in fact "loaded", which explains why the documentation is available, but for some reason everything else is unavailable.

**Hack solution.** Running `:call vimtex#init()` makes vimtex work just fine.

### Minimal working example

```tex
\documentclass{minimal}
\begin{document}

Hello World!

\end{document}
```

### Minimal vimrc file

```vim
set nocompatible
filetype plugin indent on
syntax enable
```

### Minimal vim directory

```bash
$ cd ~/.vim
$ ls
pack/ vimrc
$ tree pack
pack
└── my-plugins
    ├── opt
    └── start
        └── vimtex
            ├── after
            │   └── syntax
            │       └── tex.vim
            ├── autoload
            ...
```

**Comment by lervag** [comment]
I don't think this is related to #411. I've never attempted to use the built in package manager stuff from Vim 8. I'll test.

I'm curious: Does it work if you use the "standard" method with the "normal" package managers? You can test by adding `vimtex` to `.vim/bundle` directory, then use my "standard" minimal vimrc file:

```vim
set nocompatible

" Load Vimtex
let &rtp  = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'

" Load other plugins, if necessary
" let &rtp = '~/path/to/other/plugin,' . &rtp

filetype plugin indent on
syntax enable

" Vimtex options go here
```

**Comment by brianbuccola** [comment]
Yes, that works. (So does using pathogen.) It's strange that manually setting `'runtimepath'` like that works fine, but letting vim's package management do it does not work (especially because it works fine for every other plugin). Here is what vim's documentation says (`:h packages`):

```
When Vim starts up, after processing your .vimrc, it scans all directories in
'packpath' for plugins under the "pack/*/start" directory.  First all those
directories are added to 'runtimepath'.  Then all the plugins are loaded.
```

I have no idea what vim does "under the hood", so I don't know what the difference is between (i) how vim natively adds plugins to `'runtimepath'` and "loads" plugins and (ii) manually setting `'runtimepath'` in the "standard" way (as with your minimal example, or with pathogen).

**Comment by lervag** [comment]
Strange. I tried to reproduce your issue, but couldn't. It worked fine here. I'm on Vim version 8.0.13. When I do `echo join(split(&runtimepath, ','), "\n")`, the result is:

```
/home/lervag/.vim
/home/lervag/.vim/pack/my-plugins/start/vimtex
/usr/share/vim/vimfiles
/usr/share/vim/vim80
/usr/share/vim/vimfiles/after
/home/lervag/.vim/after
/home/lervag/.vim/pack/my-plugins/start/vimtex/after
```

Perhaps you are on an older version that has some kind of bug that has been fixed?

**Comment by brianbuccola** [comment]
I'm on the same version of vim, and doing `echo join(split(&runtimepath, ',')), "\n")` gives me the same result as you, but as I mentioned in my original point, the problem is not setting the runtimepath; the problem is that none of the commands, etc. are actually available.

Did you try `:Vim<tab>` and see if the commands are available, or try doing `cie` inside a latex environment? For me these things don't work.

**Comment by lervag** [comment]
Yes. With the current minimal example that you've provided, I get the following result when doing `:Vimtex<c-d>`:

```
:Vimtex<c-d>
VimtexClean            VimtexErrors           VimtexStop
VimtexCompile          VimtexImapsList        VimtexStopAll
VimtexCompileOutput    VimtexInfo             VimtexTocOpen
VimtexCompileSS        VimtexLabelsOpen       VimtexTocToggle
VimtexCompileSelected  VimtexLabelsToggle     VimtexToggleMain
VimtexCompileToggle    VimtexLacheck          VimtexView
VimtexCountLetters     VimtexReload
VimtexCountWords       VimtexStatus
```

As you see, all commands are available. The mappings work. And the command `cie` work as expected.

**Comment by brianbuccola** [comment]
Hmm... OK. I'll troubleshoot on my end then. Thanks for the help and also for the great plugin.

**Comment by lervag** [comment]
No problem. Don't hesitate to open a new issue or reopen/continue this discussion if you are able to pinpoint what your problem is.

**Comment by wbthomason** [comment]
Sorry to necro this thread, but I may have found (part) of the problem.

I was having the same issue as @brianbuccola, found that @lervag's MVE also worked for me, etc.

By paring down my own .vimrc, I found the following minimal non-working configuration (which uses vim-plug):

```viml
set nocompatible
call plug#begin()
Plug 'lervag/vimtex'
call plug#end()
filetype plugin indent off
```

This configuration can be made to work by moving the last line, as so:

```viml
set nocompatible
filetype plugin indent off
call plug#begin()
Plug 'lervag/vimtex'
call plug#end()
```

I am not certain if this points to an error on the side of vimtex or vim-plug, or even if it is expected behavior (though I do not think that it is). I'd be glad to try and help more, if this is an issue that others can replicate.

Off-topic: @lervag, thanks for the awesome plugin!

**Comment by lervag** [comment]
@wbthomason Thanks! Regarding your comment on the configuration: The first minimal vimrc example is clearly broken, as it turns off all filetype plugins. When using vim-plug, it is _not_ necessary to include the `filetype plugin indent off` line at all. Thus your configuration can be reduced to:

```vim
call plug#begin()
Plug 'lervag/vimtex'
call plug#end()
```

Note, the `set nocompatible` is also removed. It is only necessary if you source the file with `vim -u vimrc` or similar. However, in this case, you can do `vim -Nu vimrc` instead, which will insist on `nocompatible`.

Bottom line: This is no issue in vimtex, nor in vim-plug. Instead, it is an issue with your configuration.

**Comment by wbthomason** [comment]
Thanks for the feedback! Yes, `set nocompatible` was in there because I was using `vim -u` for my testing; I should've removed it.

`filetype plugin indent off` is in there for another reason, unrelated to vim-plug, vimtex, etc. I left it in as it seemed to exhibit the odd behavior mentioned in this thread, where the help docs for vimtex were loaded, vimtex was in the RTP, etc., but no commands or other features of vimtex were available. I didn't realize that this line would obviously cause this problem, as I thought that the way I was loading vimtex through vim-plug made vimtex not a filetype plugin, as evinced by its presence in the RTP.

Either way, your reduced config doesn't display the problem, so clearly my belief was incorrect. Thanks for the rapid and detailed response!

**Comment by lervag** [comment]
No problem! Don't hesitate to open issues if something doesn't work or if you have other questions.

**Comment by brianbuccola** [comment]
I've figured it out (kind of): the problem is that `call vimtex#init()` is not being called, and the reason is because of this block in `vimtex/ftplugin/tex.vim`:

```viml
if exists('b:did_ftplugin')
  finish
endif
let b:did_ftplugin = 1
```

If I remove this block, then vimtex loads perfectly when located in `~/.vim/pack/my-plugins/start/vimtex`. (See my very first post about how vim's native package management structure works.)

Obviously, what's happening is that `let b:did_ftplugin = 1` is being set somewhere, and when vimtex sees this, it finishes instead of calling `vimtex#init()`.

It turns out that `let b:did_ftplugin = 1` is being set by `/usr/share/vim80/ftplugin/initex.vim` (via `.../tex.vim` and `.../plaintex.vim`). If I keep the `if`-block in vimtex's `tex.vim` and instead remove the line

```viml
let b:did_ftplugin = 1
```

from `initex.vim`, then, once again, vimtex loads perfectly.

**tl;dr** `initex.vim` sets `b:did_ftplugin = 1`, which prevents vimtex's `tex.vim` from executing `call vimtex#init()`. Removing the relevant line from `initex.vim`, or removing the `if`-block from vimtex's `tex.vim`, solves the problem.

**Issue.** Neither of the above "solutions" feel like real solutions: both pieces of code are presumably there for a good reason. I don't know enough about vim's infrastructure to say why, or whether these things can be removed.

**Question.** Any idea how best to deal with all of this?

**Sub-issue.** Another weird issue is that I don't understand why this is a problem for vim's new native package manager, but not for pathogen or vim-plug.

**Comment by lervag** [comment]
@brianbuccola When you add vimtex through the new package system in the `start` directory, it should be the first one sourced. This works as expected for me, but apparently not for you. Thus it seems a different `tex.vim` ftplugin is loaded that sets `b:did_ftplugin`.

With the original minimal example, could you open a `tex` file and then do `:scriptnames` and copy the results here? It should show the order in which the different plugins are loaded.

Note: I am pretty sure the problem here is that you, for some reason, get vimtex loaded after the internal tex plugin. I.e., the internal ftplugin path is before the vimtex path in your `runtimepath` setting.

**Comment by brianbuccola** [comment]
Yes, you're right, and I've figured out why: it's because my `~/.vim` directory is actually a symlink pointing to `~/repos/dotfiles/vim`. (I keep my vim directory version-controlled on GitHub along with other dotfiles and directories.)

When I run `:scriptnames` with the MWE from above, here is what I see:

```
1: ~/.vimtex.vimrc
2: /usr/share/vim/vim80/filetype.vim
3: /usr/share/vim/vim80/ftplugin.vim
4: /usr/share/vim/vim80/indent.vim
5: /usr/share/vim/vim80/syntax/syntax.vim
6: /usr/share/vim/vim80/syntax/synload.vim
7: /usr/share/vim/vim80/syntax/syncolor.vim
8: /usr/share/vim/vim80/plugin/getscriptPlugin.vim
9: /usr/share/vim/vim80/plugin/gzip.vim
10: /usr/share/vim/vim80/plugin/logiPat.vim
11: /usr/share/vim/vim80/plugin/manpager.vim
12: /usr/share/vim/vim80/plugin/matchparen.vim
13: /usr/share/vim/vim80/plugin/netrwPlugin.vim
14: /usr/share/vim/vim80/plugin/rrhelper.vim
15: /usr/share/vim/vim80/plugin/spellfile.vim
16: /usr/share/vim/vim80/plugin/tarPlugin.vim
17: /usr/share/vim/vim80/plugin/tohtml.vim
18: /usr/share/vim/vim80/plugin/vimballPlugin.vim
19: /usr/share/vim/vim80/plugin/zipPlugin.vim
20: /usr/share/vim/vim80/ftplugin/tex.vim
21: /usr/share/vim/vim80/ftplugin/plaintex.vim
22: /usr/share/vim/vim80/ftplugin/initex.vim
23: ~/repos/dotfiles/vim/pack/my-plugins/start/vimtex/ftplugin/tex.vim
24: /usr/share/vim/vim80/indent/tex.vim
25: ~/repos/dotfiles/vim/pack/my-plugins/start/vimtex/indent/tex.vim
26: /usr/share/vim/vim80/syntax/tex.vim
27: ~/repos/dotfiles/vim/pack/my-plugins/start/vimtex/after/syntax/tex.vim
28: /usr/share/vim/vim80/syntax/dot.vim
29: /usr/share/vim/vim80/syntax/lua.vim
30: /usr/share/vim/vim80/syntax/gnuplot.vim
```

As you guessed, the internal `ftplugin/tex.vim` (#20) is being run before vimtex's (#23). (Same goes for the internal `indent/tex.vim` and `syntax/tex.vim`.)

After I rename my `~/.vim` symlink to `~/.vim.bak` and move the entire `~/repos/dotfiles/vim` directory to `~./vim`, then everything works perfectly. Here is what I see when I run `:scriptnames`:

```
1: ~/.vimtex.vimrc
2: /usr/share/vim/vim80/filetype.vim
3: /usr/share/vim/vim80/ftplugin.vim
4: /usr/share/vim/vim80/indent.vim
5: /usr/share/vim/vim80/syntax/syntax.vim
6: /usr/share/vim/vim80/syntax/synload.vim
7: /usr/share/vim/vim80/syntax/syncolor.vim
8: /usr/share/vim/vim80/plugin/getscriptPlugin.vim
9: /usr/share/vim/vim80/plugin/gzip.vim
10: /usr/share/vim/vim80/plugin/logiPat.vim
11: /usr/share/vim/vim80/plugin/manpager.vim
12: /usr/share/vim/vim80/plugin/matchparen.vim
13: /usr/share/vim/vim80/plugin/netrwPlugin.vim
14: /usr/share/vim/vim80/plugin/rrhelper.vim
15: /usr/share/vim/vim80/plugin/spellfile.vim
16: /usr/share/vim/vim80/plugin/tarPlugin.vim
17: /usr/share/vim/vim80/plugin/tohtml.vim
18: /usr/share/vim/vim80/plugin/vimballPlugin.vim
19: /usr/share/vim/vim80/plugin/zipPlugin.vim
20: ~/.vim/pack/my-plugins/start/vimtex/ftplugin/tex.vim
21: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex.vim
22: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/cmd.vim
23: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/complete.vim
24: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/util.vim
25: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/debug.vim
26: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/delim.vim
27: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/echo.vim
28: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/env.vim
29: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/fold.vim
30: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/format.vim
31: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/imaps.vim
32: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/index.vim
33: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/labels.vim
34: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/latexmk.vim
35: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/motion.vim
36: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/parser.vim
37: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/text_obj.vim
38: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/toc.vim
39: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/view.vim
40: ~/.vim/pack/my-plugins/start/vimtex/compiler/latexmk.vim
41: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/view/general.vim
42: ~/.vim/pack/my-plugins/start/vimtex/autoload/vimtex/view/common.vim
43: /usr/share/vim/vim80/ftplugin/tex.vim
44: ~/.vim/pack/my-plugins/start/vimtex/indent/tex.vim
45: /usr/share/vim/vim80/indent/tex.vim
46: /usr/share/vim/vim80/syntax/tex.vim
47: ~/.vim/pack/my-plugins/start/vimtex/after/syntax/tex.vim
48: /usr/share/vim/vim80/syntax/dot.vim
49: /usr/share/vim/vim80/syntax/lua.vim
50: /usr/share/vim/vim80/syntax/gnuplot.vim
```

Now, vimtex's `ftplugin/tex.vim` gets loaded before the internal one, which therefore loads the entire plugin.

It appears, then, that vim loads internal files before following symlinks. I'm not sure if there's a way to fix this, but anyway, this is not your problem. ;) Many thanks once again!

**Comment by lervag** [comment]
Ah, that explains it. Great! I was not aware that this would happen if you symlinked, so I learned something as well.

Personally, I keep my dotvim files in a separate repo that I clone directly to `.vim`. No need to symlink anything. Perhaps that would work for you as well?

**Comment by brianbuccola** [comment]
Yeah, I might just do that. In the meantime, adding this line to my vimrc also fixes the problem (I edited my comment above just before you replied):

```viml
set runtimepath=$HOME/repos/dotfiles/vim,$VIMRUNTIME
```

**Comment by Rmano** [comment]
Hmmm... this should be some other thing, because I have: 
```
[romano:~] 1 % ls -l .vim .vimrc
lrwxrwxrwx 1 romano romano 19 feb 10  2015 .vim -> lib/dotfiles/dotvim
lrwxrwxrwx 1 romano romano 18 feb 10  2015 .vimrc -> lib/dotfiles/vimrc
```
and all works ok, with my dotfiles loaded *before* the system ones as it should be.

**Comment by brianbuccola** [comment]
@Rmano: Just to be sure, are you using vim's native plugin manager, or are you using pathogen/vim-plug? Symlinks work fine for me with pathogen/vim-plugin, but for some reason not with the native package manager (see my first post).

**Comment by Rmano** [comment]
@brianbuccola  ah --- that must be it. I use vim-plug:  https://github.com/junegunn/vim-plug
